pkg/game/board.go

```diff
@@ -172,11 +186,119 @@ func (board *board) isPositionValid(tile elements.PlacedTile) bool {
 	return true
 }
 
-//revive:disable-next-line:unused-parameter Until the TODO is finished.
 func (board *board) CanBePlaced(tile elements.PlacedTile) bool {
-	// TODO for future tasks:
-	// - implement generation of legal moves
-	// - to be implemented after #18, #19 and #20 to avoid code duplication
+	if !board.isPositionValid(tile) {
+		return false
+	}
+	if !slices.Contains(board.placeablePositions, tile.Position) {
+		return false
+	}
+
+	meepleCount := 0
+	featuresWithMeeples := map[feature.Type]struct{}{}
+	for _, feat := range tile.Features {
+		if feat.Meeple.Type != elements.NoneMeeple {
+			// Depending on use cases for this method, meeple counting may not be
+			// strictly necessary.
+			// One example where it's unnecessary is game.PlayTurn() which calls
+			// player.PlaceTile() and that already validates meeple count
+			// for other reasons.
+			meepleCount++
+			featuresWithMeeples[feat.FeatureType] = struct{}{}
+			if meepleCount > 1 {
+				return false
+			}
+		}
+	}
+
+	for feat := range featuresWithMeeples {
+		if !canBePlacedFunctions[feat](board, tile) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (board *board) cityCanBePlaced(tile elements.PlacedTile) bool {
+	return board.cityManager.CanBePlaced(tile)
+}
+
+func (board *board) fieldCanBePlaced(tile elements.PlacedTile) bool {
+	// While placing a tile, the player can only put a single meeple on a field.
+	// This means we don't have to care about whether our field expands into
+	// a different feature on our tile - we will only expand the feature
+	// if it has a meeple and that only happens once.
+	for _, feat := range tile.Features {
+		if feat.FeatureType != feature.Field || feat.Meeple.Type == elements.NoneMeeple {
+			continue
+		}
```

Comment:
- Since we only call fieldCanBePlaced on features with a meeple, checking if meeple exists is not necessary here
- Also, instead of iterating over all features on the tile and checking if the feature type is field, we could iterate over tile.GetFeaturesOfType(feature.Field) (although that also iterates through all features, so the end result would be slower in that case, so I'm not sure about this one)

Reply:
> * Since we only call fieldCanBePlaced on features with a meeple, checking if meeple exists is not necessary here

It is necessary since only one of the field features will have a meeple and we don't know which without checking (though maybe that should just be passed to the function).

Reply:
Right, I didn't think about that

Reply:
I updated all of the functions to accept the feature and updated their implementation to use that fact.